modules/vault-cluster/main.tf

```diff
@@ -167,6 +170,26 @@ resource "aws_iam_role" "instance_role" {
   }
 }
 
+resource "aws_iam_role" "named_instance_role" {
+  # If the iam_role_name variable is set, use its value for this resource's name argument.
+  count               = "${var.iam_role_name != "" ? 1 : 0}"
+  name                = "${var.iam_role_name}"
+  assume_role_policy  = "${data.aws_iam_policy_document.instance_role.json}"
+
+  # aws_iam_instance_profile.instance_profile in this module sets create_before_destroy to true, which means
+  # everything it depends on, including this resource, must set it as well, or you'll get cyclic dependency errors
+  # when you try to do a terraform destroy.
+  lifecycle {
+    create_before_destroy = true
+  }
+}
+
+data "aws_iam_role" "instance_role" {
+  # The join() hack is required because currently the ternary operator evaluates the expressions on both branches of the
+  # condition before returning a value. This is tracked upstream: https://github.com/hashicorp/hil/issues/50
+  name = "${var.iam_role_name == "" ? join(" ", aws_iam_role.prefixed_instance_role.*.name) : var.iam_role_name }"
+}
```

Comment:
Hm, I suspect this will be slightly problematic, as it doesn't explicitly depend on the `aws_iam_role.named_instance_role` resource, so this data source may try to fetch that instance role before it's actually created. Also, the extra data source (and extra API call) seems unnecessary. I think you can replace it with a `template_file` instead:

```hcl
data "template_file" "instance_role_name" {
  template = "${element(concat(aws_iam_role.prefixed_instance_role.*.name, aws_aim_role.named_instance_role.*.name), 0)}"
}
```

You can then use `data.template_file.instance_role_name.rendered` elsewhere in the code.

Reply:
Fwiw your suspicions are confirmed, a first-run of an apply with this update would always lead to:

```
* module.vault_cluster.aws_iam_instance_profile.instance_profile: Resource 'data.aws_iam_role.instance_role' not found for variable 'data.aws_iam_role.instance_role.name'
```

I suppose that is why all the test harnessing in these modules actually goes out and creates real resources. 😄  

In hindsight, the pattern / kludge I was using initially is for cases where you want to optionally create a resource if someone isn't bringing their own *pre existing* resource identified in via a var.